### PR TITLE
Frustum Culling

### PR DESCRIPTION
**Pull request by super626**

### CODE REVIEW

File: cocos/2d/CCCamera.cpp [hunk 1]
```diff
@@ -248,6 +253,26 @@ void Camera::unproject(const Size& viewport, Vec3* src, Vec3* dst) const
     dst->set(screen.x, screen.y, screen.z);
 }
 
+void Camera::enableFrustumCull(bool bEnalbe, bool bClipZ)
```
Comment (minggo) [inline cocos/2d/CCCamera.cpp]: bEnable -> enable
bClipZ -> clipZ

File: cocos/2d/CCCamera.h [hunk 2]
```diff
@@ -135,6 +136,16 @@ class CC_DLL Camera :public Node
...
     
+    /**
+     * Enable frustum culling
+     */
+    void enableFrustumCull(bool bEnalbe, bool bClipZ);
+    
+    /**
+     * Is this aabb visible in frustum
+     */
+    bool visibleInFrustum(const AABB& aabb)const;
```
Comment (minggo) [inline cocos/2d/CCCamera.h]: visibleInFrustum -> isVisibleInFrustum?

Comment (minggo) [inline cocos/2d/CCCamera.h]: or getIsVisibleInFrustum?
Just be the same as other functions.

File: cocos/3d/CCFrustum.cpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            point = normal.dot(obb._zAxis) > 0 ? point - obbExtentZ : point + obbExtentZ;
+
+            if (_plane[i].getSide(point) == Plane::FRONT_PLANE)
+                return true;
+        }
+    }
+    return  false;
+}
+
+void Frustum::createPlane(const Camera* pcamera)
```
Comment (minggo) [inline cocos/3d/CCFrustum.cpp]: pcamera -> camera

File: cocos/3d/CCFrustum.h [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    
+protected:
+    /**
+     * create clip plane
+     */
+    void createPlane(const Camera* pcamera);
+
+    Plane _plane[6];             // clip plane, left, right, top, bottom, near, far
+    bool _bClipZ;                // use near and far clip plane
+    bool _bInit;
```
Comment (minggo) [inline cocos/3d/CCFrustum.h]: _bInit -> initialized?
Please don't use Hungary style.

File: cocos/3d/CCPlane.h [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+#include "base/ccMacros.h"
+#include "math/CCMath.h"
+
+NS_CC_BEGIN
+
+class CC_DLL Plane
+{
+public:
+    enum POINT_SIDE
```
Comment (minggo) [inline cocos/3d/CCPlane.h]: Please use enum class.

File: cocos/3d/CCPlane.h [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    {
+       ON_PLANE,
```
Comment (minggo) [inline cocos/3d/CCPlane.h]: ON_PLANE -> IN_PLANE?

File: cocos/3d/CCPlane.h (hunk 5 above)
Comment (minggo) [inline cocos/3d/CCPlane.h]: And name of enum class is the same as normal class, so POINT_SIDE -> PointSide

File: cocos/3d/CCRay.cpp [hunk 7]
```diff
@@ -140,10 +140,26 @@ bool Ray::intersects(const OBB& obb) const
     return ray.intersects(aabb);
 }
 
+float Ray::dist(const Plane& plane) const
+{
+	float ndd = Vec3::dot(plane.getNormal(),_direction);
+	if(ndd == 0)
+		return 0.0f;
```
Comment (minggo) [inline cocos/3d/CCRay.cpp]: Please replace tab with 4 spaces.

File: cocos/3d/CCRay.cpp [hunk 8]
```diff
@@ -140,10 +140,26 @@ bool Ray::intersects(const OBB& obb) const
     return ray.intersects(aabb);
 }
 
+float Ray::dist(const Plane& plane) const
+{
+	float ndd = Vec3::dot(plane.getNormal(),_direction);
```
Comment (minggo) [inline cocos/3d/CCRay.cpp]: It is better to add a space before and after `,`.

File: cocos/3d/CCPlane.cpp [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+Plane::POINT_SIDE Plane::getSide(const Vec3& point) const
+{
+    float dist = dist2Plane(point);
+    if (dist > 0)
+        return Plane::FRONT_PLANE;
+    else if (dist < 0)
+        return Plane::BEHIND_PLANE;
+
+    return Plane::ON_PLANE;
```
Comment (minggo) [inline cocos/3d/CCPlane.cpp]: How about modify like this

``` c++
float dist = dist2Plane(point);
if (dist > 0)
    return Plane::FRONT_PLANE;
else if (dist < 0)
    return Plane::BEHIND_PLANE;
else
    return Plane::ON_PLANE;
```

It is more clear.

File: cocos/3d/CCPlane.cpp [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+ ****************************************************************************/
+
+#include "CCPlane.h"
+
+NS_CC_BEGIN
+
+Plane::Plane()
+{
+    _normal.set(0.0f, 0.0f, 1.0f);
+    _dist = 0.0f;
```
Comment (minggo) [inline cocos/3d/CCPlane.cpp]: Why not use [initialization lists](http://www.cprogramming.com/tutorial/initialization-lists-c++.html)?

File: cocos/3d/CCFrustum.cpp [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+NS_CC_BEGIN
+
+bool Frustum::initFrustum(const Camera* pCamera)
+{
+    _bInit = true;
+    createPlane(pCamera);
+    return true;
+}
+bool Frustum::isOutFrustum(const AABB& aabb) const
```
Comment (minggo) [inline cocos/3d/CCFrustum.cpp]: isOutFrustum -> isOutOfFrustum?

**Comment by minggo** [comment]
And please modify all project files.

**Comment by super626** [comment]
[ci rebuild]

**Comment by minggo** [comment]
Could you please fix conflict?

**Comment by minggo** [comment]
[ci rebuild]

**Pull request merged**